// @starknet-react/core/dist/hooks/starknetid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStarknetIdContract = exports.useAddressFromStarkName = exports.useStarkName = void 0;
const react_query_1 = require("@tanstack/react-query");
const starknet_1 = require("starknet");
const providers_1 = require("../providers");
/**
 * Hook for fetching Stark name for address.
 *
 * @remarks
 *
 * This hook fetches the stark name of the specified address.
 * It defaults to the starknet.id contract but a different contract can be
 * targetted by specifying its contract address
 * If address does not have a stark name, it will return "stark"
 *
 * @example
 * This example shows how to get the stark name of an address using the default
 * Starknet.id contract
 * ```tsx
 * function Component() {
 *   const { address } = useAccount()
 *   const { data, isLoading, isError } = useStarkName({ address })
 *
 *   if (isLoading) return <span>Loading...</span>
 *   if (isError) return <span>Error fetching name...</span>
 *   return <span>StarkName: {data}</span>
 * }
 * ```
 *
 *  @example
 * This example shows how to get the stark name of an address specifying a
 * different contract address
 * ```tsx
 * function Component() {
 *   const { address } = useAccount()
 *   const { data, isLoading, isError } = useStarkName({ address, contract: '0x1234' })
 *
 *   if (isLoading) return <span>Loading...</span>
 *   if (isError) return <span>Error fetching name...</span>
 *   return <span>StarkName: {data}</span>
 * }
 * ```
 */
function useStarkName({ address, contract }) {
    const { library } = (0, providers_1.useStarknet)();
    const { data, isLoading, isSuccess, isError, error, isFetching, isStale: isIdle, isFetched, isFetchedAfterMount, isRefetching, refetch, status, } = (0, react_query_1.useQuery)({
        queryKey: ['starkName'],
        queryFn: async () => {
            const chainId = await library.getChainId();
            const provider = new starknet_1.Provider({
                sequencer: {
                    network: chainId,
                },
            });
            const namingContract = contract ?? getStarknetIdContract(chainId);
            const result = provider.getStarkName(address, namingContract);
            if (result instanceof Error)
                throw new Error(result.message);
            return result;
        },
    });
    return {
        data,
        isLoading,
        isSuccess,
        isError,
        error,
        isIdle,
        isFetching,
        isFetched,
        isFetchedAfterMount,
        isRefetching,
        refetch,
        status,
    };
}
exports.useStarkName = useStarkName;
/**
 * Hook to get the address associated to a stark name.
 *
 * @remarks
 *
 * This hook fetches the address of the specified stark name
 * It defaults to the starknetID contract but a different contract can be targetted by specifying its address
 * If stark name does not have an associated address, it will return "0x0"
 *
 * @example
 * This example shows how to get the address associated to a stark name
 * ```tsx
 * function Component() {
 *   const { data, isLoading, isError } = useAddressFromStarkName({ name: 'vitalik.stark' })
 *
 *   if (isLoading) return <span>Loading...</span>
 *   if (isError) return <span>Error fetching address...</span>
 *   return <span>address: {data}</span>
 * }
 * ```
 */
function useAddressFromStarkName({ name, contract, }) {
    const { library } = (0, providers_1.useStarknet)();
    const { data, isLoading, isSuccess, isError, error, isFetching, isStale: isIdle, isFetched, isFetchedAfterMount, isRefetching, refetch, status, } = (0, react_query_1.useQuery)({
        queryKey: ['addressFromStarkName'],
        queryFn: async () => {
            const chainId = await library.getChainId();
            const provider = new starknet_1.Provider({
                sequencer: {
                    network: chainId,
                },
            });
            const namingContract = contract ?? getStarknetIdContract(chainId);
            const result = provider.getAddressFromStarkName(name, namingContract);
            if (result instanceof Error)
                throw new Error(result.message);
            return result;
        },
    });
    return {
        data,
        isLoading,
        isSuccess,
        isError,
        error,
        isIdle,
        isFetching,
        isFetched,
        isFetchedAfterMount,
        isRefetching,
        refetch,
        status,
    };
}
exports.useAddressFromStarkName = useAddressFromStarkName;
function getStarknetIdContract(chainId) {
    const starknetIdMainnetContract = '0x6ac597f8116f886fa1c97a23fa4e08299975ecaf6b598873ca6792b9bbfb678';
    const starknetIdTestnetContract = '0x3bab268e932d2cecd1946f100ae67ce3dff9fd234119ea2f6da57d16d29fce';
    switch (chainId) {
        case starknet_1.constants.StarknetChainId.SN_MAIN:
            return starknetIdMainnetContract;
        case starknet_1.constants.StarknetChainId.SN_GOERLI:
            return starknetIdTestnetContract;
        default:
            throw new Error('Starknet.id is not yet deployed on this network');
    }
}
exports.getStarknetIdContract = getStarknetIdContract;
